src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20230926093503_EES4467_RemoveOrphanedContentBlocksAndContentSections.sql

```diff
@@ -0,0 +1,36 @@
+-- Delete orphaned Data Blocks
+DELETE FROM ContentBlock
+WHERE ContentBlock.[Type] = 'DataBlock'
+AND NOT EXISTS (
+    SELECT 1
+    FROM Publications
+    JOIN Releases ON Releases.PublicationId = Publications.Id
+    JOIN ReleaseContentBlocks ON ReleaseContentBlocks.ReleaseId = Releases.Id
+    JOIN ContentBlock cb ON cb.Id = ReleaseContentBlocks.ContentBlockId
+    AND cb.Id = ContentBlock.Id
+);
+
+-- Delete Content Blocks that are linked to orphaned Content Sections. These Content Blocks have links to existing
+-- Content Sections but those Content Sections themselves are orphaned by virtue of having no ReleaseContentSections
+-- links.
+DELETE FROM ContentBlock
+WHERE EXISTS (
+    SELECT 1
+    FROM ContentSections
+    WHERE ContentSections.Id = ContentBlock.ContentSectionId
+)
+AND NOT EXISTS (
+    SELECT 1
+    FROM ContentSections
+    JOIN ReleaseContentSections ON ReleaseContentSections.ContentSectionId = ContentSections.Id
+    WHERE ReleaseContentSections.ContentSectionId = ContentSections.Id
+    AND ContentSections.Id = ContentBlock.ContentSectionId
+);
```

Comment:
Hey Mark, thanks for the above.  I'll address it bit by bit.

_I think this is fine, as I believe the only ContentBlocks that can be used while not being attached to a ContentSection are KeyStatisticDataBlocks._

So actually, `KeyStatisticDataBlocks` are *not* a type of ContentBlock.  They're actually just a type of `KeyStatistic` that has a *link* to a ContentBlock of type "DataBlock".  Therefore, it's actually accurate to say that the only ContentBlocks that can be used while not being attached to a ContentSection are *DataBlocks*, rather than KeyStatisticDataBlocks.

_As far as I'm aware, that's the only instance of a ContentBlock that can be used without being attached to a live ContentSection are DataBlocks that are key stats._

So a DataBlock does not just need to be used as a Key Stat in order for it to be valid without being embedded in Content.  It can also be a Featured Table.  Also, it can just be a DataBlock that was created but was never actually used - I don't think we validate against that at publishing time.

Reply:
But essentially, the main point is that DataBlocks can live without being attached to any Content, whereas all other ContentBlock types are *only* linked to a Release through its content.

Reply:
`So actually, KeyStatisticDataBlocks are not a type of ContentBlock. They're actually just a type of KeyStatistic that has a link to a ContentBlock of type "DataBlock".`

I was aware of this - I actually implemented KeyStatisticDataBlocks - but didn't communicate very well. My general concern was that ContentBlocks associated with a KeyStatisticDataBlocks could potentially be removed if they've got a ContentSectionId set. It would be weird if this was true though, as that ContentSection would have to be orphaned, so it'd probably be bad data regardless.

Doing some digging, I found the EES-3988 ticket which covers this: `Data blocks associated with a key statistic no longer need a ContentSectionId, so it can be set to null.` So I'm fairly sure I cleaned this up and all ContentBlocks with an associated KeyStatDataBlock will have a null ContentSectionId.

I guess in theory the same possibility exists for ContentBlocks associated with a featured table.

Anyway, seems like you've considered this so happy to call this good.

Reply:
We'll run some queries on each environment first!

Reply:
We did this, and found none on any environment.  See https://dfedigital.atlassian.net/browse/EES-3206 latest comments.